Brokerages/WebSocketWrapper.cs

```diff
@@ -63,17 +55,110 @@ public void Initialize(string url)
         /// <param name="data"></param>
         public void Send(string data)
         {
-            _wrapped.Send(data);
+            try
+            {
+                if (!IsOpen)
+                {
+                    Log.Trace($"WebSocketWrapper.Send(): Connection not open (IsOpen:{IsOpen}, State:{_wrapped.State}): attempting connection", true);
+
+                    Connect();
+                }
+
+                var sendBuffer = new ArraySegment<byte>(Encoding.UTF8.GetBytes(data));
+
+                _wrapped.SendAsync(sendBuffer, WebSocketMessageType.Text, true, _cancellationTokenSource.Token).SynchronouslyAwaitTask();
+            }
+            catch (Exception exception)
+            {
+                OnError(new WebSocketError(exception.Message, exception));
+            }
         }
 
         /// <summary>
         /// Wraps Connect method
         /// </summary>
         public void Connect()
         {
-            if (!IsOpen)
+            if (IsOpen)
+            {
+                return;
+            }
+
+            try
+            {
+                _wrapped.ConnectAsync(new Uri(_url), _cancellationTokenSource.Token).SynchronouslyAwaitTask();
+
+                if (IsOpen)
+                {
+                    OnOpen();
+
+                    new TaskFactory().StartNew(
```

Comment:
I think there is some room for a race condition around this, eg multiple concurrent calls to `Connect` triggering multiple tasks - We might need some lock and blocking of calling threads.
Documentation at https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.receiveasync?view=netframework-4.8 states `Exactly one send and one receive is supported on each ClientWebSocket object in parallel.`